modules/graceful/server_signals.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package graceful
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+var hookableSignals []os.Signal
+
+func init() {
+	hookableSignals = []os.Signal{
+		syscall.SIGHUP,
+		syscall.SIGUSR1,
+		syscall.SIGUSR2,
+		syscall.SIGINT,
+		syscall.SIGTERM,
+		syscall.SIGTSTP,
+	}
+}
+
+// handleSignals listens for os Signals and calls any hooked in function that the
+// user had registered with the signal.
+func (srv *Server) handleSignals() {
+	var sig os.Signal
+
+	signal.Notify(
+		srv.sigChan,
+		hookableSignals...,
+	)
+
+	pid := syscall.Getpid()
+	for {
+		sig = <-srv.sigChan
+		srv.preSignalHooks(sig)
+		switch sig {
+		case syscall.SIGHUP:
+			if setting.GracefulRestartable {
+				log.Info("PID: %d. Received SIGHUP. Forking...", pid)
+				err := srv.fork()
+				if err != nil {
+					log.Error("Error whilst forking from PID: %d : %v", pid, err)
+				}
+			} else {
+				log.Info("PID: %d. Received SIGHUP. Not set restartable. Shutting down...", pid)
+
+				srv.shutdown()
```

Comment:
Two ways of shutting down a process in Go:

1) Kill it. E.g. The hammer function
2) the Main function returns, which in the case of the web command, boils down to runWeb finishing. 

That boils down to this or similar returning:

https://github.com/go-gitea/gitea/blob/f1fdd782d57ddbd2a759dad2843ee619709a4609/cmd/web.go#L167 

I.e. this function returns

https://github.com/go-gitea/gitea/blob/c618a85158d66c08177da1a2ee0d9070209c28b3/modules/graceful/server.go#L173-L185

Which happens when the listener is closed and the waitgroup of open connections are done.

Reply:
Just to be a bit clearer, here:

```go
err := serve(srv.Listener)
```

The `serve` function is passed in to this function. 

`serve` is basically a function with an infinite for loop that accepts new connections from the listener and then processes the requests (spawning goroutines/from pools/or sequentially.) If a listener is closed then accept will return an error and that is usually just returned (and in the case of HTTP serves that's what's happens.) Otherwise the serve function never terminates.

Reply:
Thanks, `srv.wg.Wait()` is what I was looking for. I couldn't find how the web server knew about requests finishing.